storey/flow.py

```diff
@@ -440,81 +496,97 @@ def __init__(self, paths, with_header=False, build_dict=False, key_field=None, t
+                            timestamp_field = field_name_to_index[timestamp_field]
+                        timestamp_str = parsed_line[timestamp_field]
+                        if self._timestamp_format:
+                            timestamp = datetime.strptime(timestamp_str, self._timestamp_format)
                         else:
-                            timestamp = datetime.now()
-                        await self._do_downstream(Event(element, key=key, time=timestamp))
-            termination_result = await self._do_downstream(_termination_obj)
-            self._termination_future.set_result(termination_result)
-        except BaseException as ex:
-            self._ex = ex
-            self._termination_future.set_result(None)
+                            timestamp = datetime.fromisoformat(timestamp_str)
+                    else:
+                        timestamp = datetime.now()
+                    await self._do_downstream(Event(element, key=key, time=timestamp))
+        return await self._do_downstream(_termination_obj)
 
-    def _loop_thread_main(self):
-        asyncio.run(self._run_loop())
-        self._termination_q.put(self._ex)
 
-    def _raise_on_error(self, ex):
-        if ex:
-            raise FlowError('Flow execution terminated due to an error') from self._ex
+async def _aiter(iterable):
+    for x in iterable:
+        yield x
 
-    def run(self):
-        super().run()
 
-        thread = threading.Thread(target=self._loop_thread_main)
-        thread.start()
+class DataframeSource(FileSource):
+    """
+        Use pandas dataframe as input source for a flow.
+
+        :param dfs: A pandas dataframe, or dataframes, to be used as input source for the flow.
+        :type paths: pandas.DataFrame, or list of pandas.DataFrame
+        :param key_column: column to be used as key for events.
+        :type key_column: string
+        :param time_column: column to be used as time for events.
+        :type time_column: datetime
+        :param id_column: column to be used as ID for events.
+        :type id_column: string
+    """
 
-        def raise_error_or_return_termination_result():
-            self._raise_on_error(self._termination_q.get())
-            return self._termination_future.result()
+    def __init__(self, dfs, key_column=None, time_column=None, id_column=None, **kwargs):
+        super().__init__(**kwargs)
+        if not isinstance(dfs, list):
+            dfs = [dfs]
+        self._dfs = dfs
+        self._key_field = key_column
+        self._time_field = time_column
+        self._id_field = id_column
 
-        return FlowAwaiter(raise_error_or_return_termination_result)
+    async def _run_loop(self):
+        async for df in _aiter(self._dfs):
+            async for _, row in _aiter(df.iterrows()):
+                body = row.to_dict()
+                key = None
+                if self._key_field:
+                    key = body.pop(self._key_field, None)
```

Comment:
I would raise KeyError rather than silencing it.
its a user error in he should be aware of that, he might have misspelled the column name or whatever